### PR TITLE
prov/efa: check if hmem device is initialized, add check to mr regattr

### PR DESCRIPTION
**Pull request by rwespetal**
Instead of just relying on the sysfs file, validate that hmem support
was initialized for that particular device during core initialization,
and don't report FI_HMEM support if support is not available.

Also add some checks fo the mr regattr paths to validate whether support
exists. This isn't strictly necessary, but may help users when the
provider supports multiple devices. The current FI_HMEM flag doesn't
provide information on which devices are supported.

This doesn't contain changes around whether p2p support is available, and
falling back to copies when it is not, so p2p is still required for EFA +
FI_HMEM at this point.

Signed-off-by: Robert Wespetal <wesper@amazon.com>

Test description: Verified `fi_rdm_pingpong` still works on p4 with both cuda and host buffers. Also made some changes to libfabric/fabtests/unloaded drivers to manually test the other paths in the initialization and registration paths here.

### CODE REVIEW

**Review by bwbarrett — APPROVED** [review]

File: prov/efa/src/efa_mr.c [hunk 1]
```diff
@@ -57,6 +57,62 @@ static struct fi_ops efa_mr_cache_ops = {
...
+	if (attr->iface == FI_HMEM_SYSTEM) {
+		efa_mr->peer.iface = FI_HMEM_SYSTEM;
+		return FI_SUCCESS;
+	} else if (efa_mr->domain->util_domain.info_domain_caps & FI_HMEM) {
+		/*
+		 * Skipping the domain type check above is okay here since
+		 * util_domain is at the beginning of both efa_domain and
+		 * rxr_domain.
+		 */
+		if (efa_mr->domain->ctx->supported_hmem_devs & attr->iface) {
```
Comment (wzamazon) [inline prov/efa/src/efa_mr.c:88]: I do not think we can use `attr->face` as flag, because it is an enum.

For example, FI_HMEM_ZE's value is 3, which is FI_HMEM_CUDA |  FI_HMEM_ROCR.

Also, can we just call `ofi_hmem_is_initialized(attr->iface)` here?

Comment (rwespetal) [reply prov/efa/src/efa_mr.c:88]: Yep, you're right, I'll fix this.

I think we still need internal state in our provider to decide whether it supports that device type. That device type could be supported by Libfabric, but not the provider.

Comment (rwespetal) [reply prov/efa/src/efa_mr.c]: I thought about it some more and removed the flag for now. We'll want to store additional information, like if p2p is supported/enabled, and it's more appropriate to store that in the endpoint.

For now `domain_caps & FI_HMEM` validates what we need and we can extend this later.

File: prov/efa/src/efa_mr.c [hunk 2]
```diff
@@ -57,6 +57,58 @@ static struct fi_ops efa_mr_cache_ops = {
...
+ * Check if FI_HMEM is enabled for the domain, validate whether the specific
+ * device type requested is currently supported by the provider, and update the
+ * efa_mr structure based on the attributes requested by the user.
+ *
+ * @params[in]	efa_mr	efa_mr structure to be updated
+ * @params[in]	attr	fi_mr_attr from the user's registration call
+ *
+ * @return FI_SUCCESS or negative FI error code
+ */
+static int efa_ep_mr_hmem_setup(struct efa_mr *efa_mr,
```
Comment (wzamazon) [inline prov/efa/src/efa_mr.c:72]: nit: this function is not related to ep, so the name can be just `efa_mr_hmem_setup`.

Comment (rwespetal) [reply prov/efa/src/efa_mr.c]: Good point, fixed.

**Review by wzamazon — APPROVED** [review]
